snuba/query/parser/validation/functions.py

```diff
@@ -40,8 +41,38 @@ def validate(self, exp: Expression, data_source: DataSource) -> None:
 
         entity_validators: Mapping[str, FunctionCallValidator] = {}
         data_model: Optional[ColumnSet] = None
+        # If the data_source is a JoinClause it will have mutiple
+        # entities so we want to make sure we validate them all.
+        entities: List[Entity] = []
+
         if isinstance(data_source, QueryEntity):
             entity = get_entity(data_source.key)
+            entities.append(entity)
+            data_model = entity.get_data_model()
+
+        elif isinstance(data_source, JoinClause):
+            alias_map = data_source.get_alias_node_map()
+            for _, node in alias_map.items():
+                assert isinstance(node.data_source, QueryEntity)  # mypy
+                entities.append(get_entity(node.data_source.key))
+            data_model = data_source.get_columns()
+
+        elif isinstance(data_source, ProcessableQuery):
+            entity = get_entity(data_source.get_from_clause().key)
```

Comment:
You don't need to do this, `_post_process` will already recurse on the query. If you have this here you're going to validate twice.

Reply:
I think this is required actually, but not in this form `isinstance(data_source, ProcessableQuery)` is not broad enoguh.
In a hypothetical query with a hypothetical function call like
```
MATCH (
     MATCH entity
     SELECT .......whatever ....
)
SELECT shutdown_system_and_format_hdd(), event_id
```

We do want to validate the shutdown function call against the list of allowed/denied functions.
In that case data_source would be a `CompositeQuery` if I get this right.